Refactor and clean up the decode/encode_pg_control functions

hlinnaka opened the pull request:
See commit messages.

This tries to document the transmute issues (https://github.com/zenithdb/zenith/issues/207) in code comments. How to solve that is a separate discussion, but let's at least add warnings to the code about it, until it's fixed.

hlinnaka commented on postgres_ffi/build.rs:
```diff
@@ -11,16 +11,19 @@ fn main() {
-        // bindings for.
+        // All the needed PostgreSQL headers are included from 'pg_control_ffi.h'
         .header("pg_control_ffi.h")
+
```
I added these newlines for clarity, to separate the different kinds of options. But `cargo fmt` insists on removing them again :-(. Any ideas?

funbringer replied:
Idk, what about this?

```
Builder::default()
    // section one
    .foo()
    .bar()
    .baz()
    //
    // section two
    .foo()
    .bar()
    .baz()
    //
    // section three
    .foo()
    .bar()
    .baz();
```

hlinnaka replied:
did it that way

ericseppanen commented:
In #208 I was able to tidy up the crc code: [link](https://github.com/zenithdb/zenith/pull/208/files#diff-526683ea09888f69e04822fada89ba4d86b1bf4dd6c99e479d9e2af28b68c016R144). No `OFFSETOF_CRC` const needed.

Also, encoding and decoding functions should be implemented against the struct; no need for them to be free-floating functions.

funbringer commented on postgres_ffi/src/controlfile_utils.rs:
```diff
@@ -0,0 +1,105 @@
+/// Convert a struct representing a Postgres control file into raw bytes.
+///
+/// The CRC is recomputed to match the contents of the fields.
+pub fn encode_pg_control(controlfile: ControlFileData) -> Bytes {
```
I also feel that those two free functions belong to the `impl ControlFileData` above.

Then they would become

```rust
impl ControlFileData {
    pub fn decode(buf: &[u8]) -> anyhow::Result<Self>;  // becomes ControlFileData::decode at call site
    pub fn encode(&self) -> Bytes;  // e.g. ControlFileData::new().encode()
}
```

Note how we don't have to consume ControlFileData in encode.

funbringer commented on postgres_ffi/src/controlfile_utils.rs:
```diff
@@ -0,0 +1,105 @@
+    //
+    // Use `transmute` to reinterpret struct as raw bytes.
+    //
+    // FIXME: This triggers undefined behavior, because the contents
```
Just wanted to say that I appreciate thoughtful comments like this one.

hlinnaka commented:
> In #208 I was able to tidy up the crc code: [link](https://github.com/zenithdb/zenith/pull/208/files#diff-526683ea09888f69e04822fada89ba4d86b1bf4dd6c99e479d9e2af28b68c016R144). No `OFFSETOF_CRC` const needed.

Ah, cool, I missed that part of #208. Cherry-picked that here.

hlinnaka commented:
Thanks for the review so far. Can you please take another look now, this has changed quite a bit since the first version. Notably:

- I adopted the `memoffset::offset_of!` replacement for C offsetof()
- I turned encode/decode into method in ControlFileData
- I added a comment at the top of controlfile_utils.rs to give an introduction to what the control file is
- I removed the `ControlFileData::new()` constructor, it was unused

funbringer commented on postgres_ffi/src/controlfile_utils.rs:
```diff
@@ -0,0 +1,125 @@
+        // another tenant's data.
+        //
+        // See https://github.com/zenithdb/zenith/issues/207 for discussion.
+        let b: [u8; SIZEOF_CONTROLDATA] =
+            unsafe { std::mem::transmute::<ControlFileData, [u8; SIZEOF_CONTROLDATA]>(*self) };
```
Another way to do it:

```rust
let b: [u8; SIZEOF_CONTROLDATA] = unsafe {
    std::mem::transmute_copy(self)
};
```

Unfortunately, this function doesn't account for different src/dst sizes (which is unsafe squared), so I doubt that we should use it.

As an aside: rust is able to infer types in such cases, which means that we don't have to write the bulky `<ControlFileData, [u8; SIZEOF_CONTROLDATA]>`. Disregard this if you feel more comfortable with extra explicit types in unsafe blocks.

funbringer reviewed (approved):
LGTM, but I'd like to hear from Eric.

One more thing: I wonder if we should rename `encode -> to_bytes` and `decode -> from_bytes` to be more in line with [rust's naming conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv), but I have no strong opinions on this subject.
